src/vs/workbench/contrib/preferences/browser/settingsList.ts

```diff
@@ -0,0 +1,437 @@
+	dispose() {
+		[...this.usedPool.values(), ...this.freePool.values()].forEach(ts => ts.forEach(({ template }) => {
+			template.toDispose.dispose();
+			template.elementDisposables.dispose();
+		}));
+
+		this.usedPool.clear();
+		this.freePool.clear();
+		this.pageDisposables.dispose();
+
+		super.dispose();
+	}
+
+	constructor(
+		private container: HTMLElement,
+		viewState: ISettingsEditorViewState,
+		renderers: ITreeRenderer<SettingsTreeElement, never, any>[],
+		@IThemeService private themeService: IThemeService,
+		@IConfigurationService configurationService: IConfigurationService,
+		@IInstantiationService instantiationService: IInstantiationService,
+	) {
+		super();
+
+		container.setAttribute('tabindex', '-1');
+		container.setAttribute('role', 'form');
+		container.setAttribute('aria-label', localize('settings', "Settings"));
+		container.classList.add('settings-editor-tree');
+
+		renderers.forEach(renderer => this.templateToRenderer.set(renderer.templateId, renderer));
+
+		this.searchFilter = element => instantiationService.createInstance(SettingsTreeFilter, viewState).filter(element, null as any);
+
+		this._register(registerThemingParticipant(themingParticipant));
+	}
+
+	getHTMLElement(): HTMLElement {
+		return this.container;
+	}
+
+	refresh(rootGroup: SettingsTreeGroupElement): void {
+		let scrollToPage: number | undefined = 1;
+
+		if (isDefined(this.currentView)) {
+			const refreshedGroup = findGroup(rootGroup, this.currentView.group.id);
+
+			if (isDefined(refreshedGroup)) {
+				this.currentView = this.getSettingsFromGroup(refreshedGroup);
+				scrollToPage = undefined;
+			} else {
+				this.currentView = undefined;
+			}
+		}
+
+		this.currentView = this.currentView ?? this.getSettingsFromGroup(rootGroup);
+		this.paginator.setSettings(this.currentView.settings, scrollToPage);
+	}
+
+	render(group: SettingsTreeGroupElement): void {
+		this.currentView = this.getSettingsFromGroup(group);
+		this.paginator.setSettings(this.currentView.settings, 1);
+	}
+
+	jumpToSetting(element: SettingsTreeSettingElement): void {
+		if (isUndefinedOrNull(element.parent)) {
+			return;
+		}
+
+		this.currentView = this.getSettingsFromGroup(element.parent);
+		this.currentView.focusedSetting = element;
+
+		const idxInView = this.currentView.settings.findIndex(setting => setting.id === element.id);
+		if (idxInView < 0) {
+			return;
+		}
+
+		// idxInView + 1: if idx is 0, then ceil is 0. But page numbers start at 1.
+		const scrollToPage = Math.ceil((idxInView + 1) / this.paginator.PAGE_SIZE);
+		this.paginator.setSettings(this.currentView.settings, scrollToPage);
+	}
+
+	private renderPage(shouldScroll: boolean): void {
+		DOM.clearNode(this.container);
+		this.pageDisposables.clear();
+
+		// Transfer all used items to the free pool
+		for (const [templateId, usedItems] of this.usedPool.entries()) {
+			const freeItems = this.freePool.get(templateId) ?? [];
+			usedItems.forEach(item => item.template.elementDisposables.clear());
```

Comment:
I think you should call `disposeElement` on the renderer instead. This level shouldn't know anything about what's inside a template.